planner/core/exhaust_physical_plans_test.go

```diff
@@ -277,22 +332,150 @@ func TestIndexJoinAnalyzeLookUpFilters(t *testing.T) {
 		},
 	}
 	for i, tt := range tests {
-		pushed, err := rewriteSimpleExpr(ctx, tt.pushedDownConds, dsSchema, dsNames)
-		require.NoError(t, err)
-		dataSourceNode.pushedDownConds = pushed
-		others, err := rewriteSimpleExpr(ctx, tt.otherConds, joinNode.schema, joinColNames)
-		require.NoError(t, err)
-		joinNode.OtherConditions = others
-		helper := &indexJoinBuildHelper{join: joinNode, lastColManager: nil, innerPlan: dataSourceNode}
-		_, err = helper.analyzeLookUpFilters(path, dataSourceNode, tt.innerKeys, tt.innerKeys, false)
-		if helper.chosenRanges == nil {
-			helper.chosenRanges = ranger.Ranges{}
+		testAnalyzeLookUpFilters(t, indexJoinCtx, &tt, fmt.Sprintf("test case: %v", i))
+	}
+}
+
+func checkRangeFallbackAndReset(t *testing.T, ctx sessionctx.Context, expectedRangeFallback bool) {
+	require.Equal(t, expectedRangeFallback, ctx.GetSessionVars().StmtCtx.RangeFallback)
+	ctx.GetSessionVars().StmtCtx.RangeFallback = false
+}
+
+func TestRangeFallbackForAnalyzeLookUpFilters(t *testing.T) {
+	ijCtx := prepareForAnalyzeLookUpFilters()
+	ctx := ijCtx.dataSourceNode.ctx
+	dsSchema := ijCtx.dataSourceNode.schema
+
+	type testOutput struct {
+		ranges         string
+		idxOff2KeyOff  string
+		accesses       string
+		remained       string
+		compareFilters string
+	}
+
+	tests := []struct {
+		innerKeys       []*expression.Column
+		pushedDownConds string
+		otherConds      string
+		outputs         []testOutput
+	}{
+		{
+			innerKeys:       []*expression.Column{dsSchema.Columns[1], dsSchema.Columns[3]},
+			pushedDownConds: "a in (1, 3) and c in ('aaa', 'bbb')",
+			otherConds:      "",
+			outputs: []testOutput{
+				{
+					ranges:         "[[1 NULL \"aa\" NULL,1 NULL \"aa\" NULL] [1 NULL \"bb\" NULL,1 NULL \"bb\" NULL] [3 NULL \"aa\" NULL,3 NULL \"aa\" NULL] [3 NULL \"bb\" NULL,3 NULL \"bb\" NULL]]",
+					idxOff2KeyOff:  "[-1 0 -1 1 -1]",
+					accesses:       "[in(Column#1, 1, 3) in(Column#3, aaa, bbb)]",
+					remained:       "[in(Column#3, aaa, bbb)]",
+					compareFilters: "<nil>",
+				},
+				{
+					ranges:         "[[1 NULL \"aa\",1 NULL \"aa\"] [1 NULL \"bb\",1 NULL \"bb\"] [3 NULL \"aa\",3 NULL \"aa\"] [3 NULL \"bb\",3 NULL \"bb\"]]",
+					idxOff2KeyOff:  "[-1 0 -1 -1 -1]",
+					accesses:       "[in(Column#1, 1, 3) in(Column#3, aaa, bbb)]",
+					remained:       "[in(Column#3, aaa, bbb)]",
+					compareFilters: "<nil>",
+				},
+				{
+					ranges:         "[[1 NULL,1 NULL] [3 NULL,3 NULL]]",
+					idxOff2KeyOff:  "[-1 0 -1 -1 -1]",
+					accesses:       "[in(Column#1, 1, 3)]",
+					remained:       "[in(Column#3, aaa, bbb)]",
+					compareFilters: "<nil>",
+				},
+				{
+					ranges:         "[]",
+					idxOff2KeyOff:  "[]",
+					accesses:       "[]",
+					remained:       "[]",
+					compareFilters: "<nil>",
```

Comment:
Does this mean we fail to choose an index join?

Reply:
Yes. If there is no key in index join inner ranges(while eq conditions may still be in inner ranges), then hash join is a better choice. In this case we give up using index join.